addons/mail/models/mail_message.py

```diff
@@ -731,30 +697,38 @@ def set_message_done(self):
         if not notifications:
             return
 
+        notifications.write({'is_read': True})
+
         # notifies changes in messages through the bus.  To minimize the number of
         # notifications, we need to group the messages depending on their channel_ids
         groups = []
         messages = notifications.mapped('mail_message_id')
-        current_channel_ids = messages[0].channel_ids
+        messages_channel = messages.filtered(lambda msg: msg.model == "mail.channel")
+        if not messages_channel:
+            return
+
+        current_channel_ids = [messages_channel[0].res_id]
         current_group = []
-        for record in messages:
-            if record.channel_ids == current_channel_ids:
+        for record in messages_channel:
+            if record.res_id == current_channel_ids:
                 current_group.append(record.id)
             else:
                 groups.append((current_group, current_channel_ids))
                 current_group = [record.id]
-                current_channel_ids = record.channel_ids
+                current_channel_ids = record.res_id
 
         groups.append((current_group, current_channel_ids))
         current_group = [record.id]
-        current_channel_ids = record.channel_ids
-
-        notifications.write({'is_read': True})
+        current_channel_ids = [record.res_id]
 
         for (msg_ids, channel_ids) in groups:
             # channel_ids in result is deprecated and will be removed in a future version
-            notification = {'type': 'mark_as_read', 'message_ids': msg_ids, 'channel_ids': [c.id for c in channel_ids], 'needaction_inbox_counter': self.env.user.partner_id.get_needaction_count()}
-            self.env['bus.bus'].sendone((self._cr.dbname, 'res.partner', partner_id.id), notification)
+            self.env['bus.bus'].sendone((self._cr.dbname, 'res.partner', partner_id.id), {
+                'type': 'mark_as_read',
+                'message_ids': msg_ids,
+                'channel_ids': channel_ids,
```

Comment:
you can remove the channel_ids key, we don't use it anymore (+ all the crap above to compute it)